text/0039-guarded-arith.md

```diff
@@ -0,0 +1,95 @@
+# Guard Arithmetic Expressions within Policies
+
+## Related issues and PRs
+
+- Reference Issues: No existing issue, but supersedes https://github.com/cedar-policy/rfcs/pull/36
+- Implementation PR(s): (Existing PR for range analysis based on schema: https://github.com/cedar-policy/cedar/pull/56)
+
+## Timeline
+
+- Started: 2023-11-07
+
+## Summary
+
+We will use a range analysis (as in https://github.com/cedar-policy/cedar/pull/56) but with the bounds given in the policy rather than in the schema.
+
+## Motivation
+
+Arithmetic with bounded integers is a common source of bugs [https://blog.research.google/2006/06/extra-extra-read-all-about-it-nearly.html]. Consider the following policy:
+
+```
+forbid(principal, action, resource) when {
+   principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms ||
+   !principal.is_mfa
+};
+```
+
+A casual reader of this policy may conclude that all actions are forbiden when `principal.mfa` is `false`. This is wrong. The lhs (`principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms`) is evaluated first and if it errors the entire policy is skipped.
+
+To avoid this class of mistakes, we want to change Cedar's validator to error on arithmetic expressions that could overflow. To pass validation, arithmetic should be guarded. For example, the above policy should be written:
+
+```
+forbid(principal, action, resource) when {
+   (principal.elevated_privileges_end_time_s < i64::MAX / 1000 &&
+    principal.elevated_privileges_end_time_s > i64::MIN / 1000 &&
+    principal.elevated_privileges_end_time_s * 1000 > context.current_time_ms) ||
+   !principal.is_mfa
+};
+```
+
+## Detailed design
+
+Every non-identity preserving arithmetic operation should be guarded by comparisons that prevent over/underflow.
+
+- We add intervals to represent the ranges of each integer.
```

Comment:
We probably want to refine what "each integer" means here. For instance, we probably don't want annotate literals with ranges.

Reply:
The RFC here is talking about the internal implementation. In analyzing the code, it will associate an interval with each integer-typed expression, and then refine that interval according to comparison operations. So consider this expression:
```
MIN_A <= principal.a && // A 
principal.a <= MAX_A && // B
... principal.a + 100 ... 
```
You start with an interval of `principal.a` as `[i64::MIN,i64::MAX]`. Then at position A you will have refined this interval to be `[MIN_A,i64::MAX]`. At position B you will have refined the interval to be `[MIN_A,MAX_A]`. Finally, based on this, the expression `principal.a + 100` will be checked to ensure that `MIN_A + 100` is not out of bounds (it isn't) and likewise that `MAX_A + 100` is not out of bounds.

A literal's interval will just have the min/max be the interval itself, e.g., the literal 5 will have the interval `[5,5]`. There are no annotations.